cranelift/peepmatic/crates/runtime/src/linear.rs

```diff
@@ -0,0 +1,219 @@
+/// result from this increment's matching operation. Each increment will
+/// basically become a state and a transition edge out of that state in the
+/// final automata.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Increment {
+    /// The matching operation to perform.
+    pub operation: MatchOp,
+
+    /// The expected result of our matching operation, that enables us to
+    /// continue to the next increment. `None` is used for wildcard-style "else"
+    /// transitions.
+    pub expected: Option<u32>,
+
+    /// Actions to perform, given that the operation resulted in the expected
+    /// value.
+    pub actions: Vec<Action>,
+}
+
+/// A matching operation to be performed on some Cranelift instruction as part
+/// of determining whether an optimization is applicable.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Deserialize, Serialize)]
+pub enum MatchOp {
+    /// Switch on the opcode of an instruction.
+    Opcode {
+        /// The path to the instruction whose opcode we're switching on.
+        path: PathId,
+    },
+
+    /// Does an instruction have a constant value?
+    IsConst {
+        /// The path to the instruction (or immediate) that we're checking
+        /// whether it is constant or not.
+        path: PathId,
+    },
+
+    /// Is the constant value a power of two?
+    IsPowerOfTwo {
+        /// The path to the instruction (or immediate) that we are checking
+        /// whether it is a constant power of two or not.
+        path: PathId,
+    },
+
+    /// Switch on the bit width of a value.
+    BitWidth {
+        /// The path to the instruction (or immediate) whose result's bit width
+        /// we are checking.
+        path: PathId,
+    },
+
+    /// Does the value fit in our target architecture's native word size?
+    FitsInNativeWord {
+        /// The path to the instruction (or immediate) whose result we are
+        /// checking whether it fits in a native word or not.
+        path: PathId,
+    },
+
+    /// Are the instructions (or immediates) at the given paths the same?
+    Eq {
+        /// The path to the first instruction (or immediate).
+        path_a: PathId,
+        /// The path to the second instruction (or immediate).
+        path_b: PathId,
+    },
+
+    /// Switch on the constant integer value of an instruction.
+    IntegerValue {
+        /// The path to the instruction.
+        path: PathId,
+    },
+
+    /// Switch on the constant boolean value of an instruction.
+    BooleanValue {
+        /// The path to the instruction.
+        path: PathId,
+    },
+
+    /// Switch on a condition code.
+    ConditionCode {
+        /// The path to the condition code.
+        path: PathId,
+    },
+
+    /// No operation. Always evaluates to `None`.
+    ///
+    /// Exceedingly rare in real optimizations; nonetheless required to support
+    /// corner cases of the DSL, such as a LHS pattern that is nothing but a
+    /// variable pattern.
+    Nop,
```

Comment:
```suggestion
    False,
```

Reply:
`Nop` is closer to `True` than `False`, but I think renaming it really clarifies anything here.